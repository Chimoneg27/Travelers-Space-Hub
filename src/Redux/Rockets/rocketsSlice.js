/* eslint-disable import/no-extraneous-dependencies */
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const rocketAPI = 'https://api.spacexdata.com/v4/rockets';

export const getRocketData = createAsyncThunk('rockets/getRocketData', async () => {
  const response = await axios.get(rocketAPI);
  return response.data;
});

const rocketReducer = createSlice({
  name: 'SpaceX Rockets',
  initialState: {
    rocketArr: [],
    isLoading: false,
    error: null,
    isDataFetched: false,
  },
  reducers: {
    addReservation: (state, action) => {
      const updatedRocketArr = state.rocketArr.map((rocketObj) => (rocketObj.id === action.payload
        ? { ...rocketObj, reserved: false } : rocketObj));
      return { ...state, rocketArr: updatedRocketArr };
    },
    cancelReservation: (state, action) => {
      const updatedRocketArr = state.rocketArr.map((rocketObj) => (rocketObj.id === action.payload
        ? { ...rocketObj, reserved: true } : rocketObj));
      return { ...state, rocketArr: updatedRocketArr };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getRocketData.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(getRocketData.fulfilled, (state, action) => {
        state.isLoading = false;
        state.rocketArr = action.payload.map((rocket) => ({
          id: rocket.id,
          name: rocket.name,
          flickr_images: rocket.flickr_images,
          description: rocket.description,
        }));
        state.isDataFetched = true;
      })
      .addCase(getRocketData.rejected, (state, action) => {
        state.isLoading = false;
        state.error = action.error;
      });
  },
});

export const { cancelReservation, addReservation } = rocketReducer.actions;
export default rocketReducer.reducer;
