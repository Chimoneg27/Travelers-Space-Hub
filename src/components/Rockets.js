import React from 'react';
import Navigation from './Navigation';

const Rockets = () => (
  <div className="rocket-page">
    <Navigation />
    <h1>Rockets</h1>
  </div>
);

export default Rockets;
