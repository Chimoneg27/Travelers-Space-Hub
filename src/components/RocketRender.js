import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { cancelReservation, addReservation } from '../Redux/Rockets/rocketsSlice';

const RocketRender = ({ rocket }) => {
  const dispatch = useDispatch();

  const handleReserve = (arg) => {
    dispatch(addReservation(arg));
  };

  const handleCancel = (arg) => {
    dispatch(cancelReservation(arg));
  };

  return (
    <div className="RocketHolder" key={rocket.id}>
      {rocket.flickr_images.length > 0 && (
      <img className="rocket-img" src={rocket.flickr_images[0]} alt={rocket.name} />
      )}
      <div className="rocket-info">

        <h3>{rocket.name}</h3>
        <p>
          <span>
            {rocket.reserved ? (<span className="Badge">Reserved</span>) : ''}
          </span>
          {rocket.description}
        </p>
        <div>
          {rocket.reserved ? (
            <button
              className="cancel-btn"
              type="button"
              onClick={() => handleReserve(rocket.id)}
            >
              Cancel Reservation
            </button>
          ) : (
            <button
              className="reserve-btn"
              type="button"
              onClick={() => handleCancel(rocket.id)}
            >
              Reserve Rocket
            </button>
          )}
        </div>
      </div>
    </div>
  );
};

RocketRender.propTypes = {
  rocket: PropTypes.oneOfType([PropTypes.object]).isRequired,
};

export default RocketRender;
