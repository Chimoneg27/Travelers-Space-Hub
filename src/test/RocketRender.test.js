import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Rockets from '../components/Rockets';
import RocketRender from '../components/RocketRender';

const mockStore = configureMockStore([]);
const store = mockStore({
  rockets: {
    rocketArr: [
      {
        id: '1',
        name: 'Rocket1',
        flickr_images: ['image1'],
        description: 'Description1',
      },
      {
        id: '2',
        name: 'Rocket2',
        flickr_images: ['image2'],
        description: 'Description2',
      },

    ],
    isDataFetched: true,
  },
});
test('Check RocketRender', async () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Rockets>
        <RocketRender />
      </Rockets>
    </Provider>,

  ).toJSON();
  expect(tree).toMatchSnapshot();
});
