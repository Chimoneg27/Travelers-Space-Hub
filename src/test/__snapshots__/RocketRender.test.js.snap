// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check RocketRender 1`] = `
<div
  className="rocket-page"
>
  <div
    className="rocket-container"
  >
    <div
      className="RocketHolder"
    >
      <img
        alt="Rocket1"
        className="rocket-img"
        src="image1"
      />
      <div
        className="rocket-info"
      >
        <h3>
          Rocket1
        </h3>
        <p>
          <span />
          Description1
        </p>
        <div>
          <button
            className="reserve-btn"
            onClick={[Function]}
            type="button"
          >
            Reserve Rocket
          </button>
        </div>
      </div>
    </div>
    <div
      className="RocketHolder"
    >
      <img
        alt="Rocket2"
        className="rocket-img"
        src="image2"
      />
      <div
        className="rocket-info"
      >
        <h3>
          Rocket2
        </h3>
        <p>
          <span />
          Description2
        </p>
        <div>
          <button
            className="reserve-btn"
            onClick={[Function]}
            type="button"
          >
            Reserve Rocket
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
